pkg/power/accelerator/qat/source/qat_utils.go

```diff
@@ -0,0 +1,201 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+
+Copyright 2023.
+*/
+
+package source
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"regexp"
+	"strconv"
+	"strings"
+)
+
+var (
+	//telemetry base path
+	teleBasePath = "/sys/devices/pci0000:%s/0000:%s:00.0/telemetry/"
+
+	//control telemetry swtich path
+	controlPath = filepath.Join(teleBasePath, "control")
+
+	//obtain device utilization data path
+	deviceDataPath = filepath.Join(teleBasePath, "device_data")
+)
+
+// getDevices obtain available qat devices and search for ID
+func getDevices() (map[string]interface{}, error) {
+	//Available devices
+	availableDev := make(map[string]interface{})
+
+	//use adf_ctl get qat devices status
+	commandText := "adf_ctl status"
+	cmd := exec.Command("bash", "-c", commandText)
+	out, err := cmd.CombinedOutput()
+	if err != nil {
+		return nil, fmt.Errorf("could not get qat status %s", err)
+	}
+	lines := strings.Split(string(out), "\n")
+
+	// Regular expression pattern, matching rows that meet the condition
+	pattern := regexp.MustCompile(`(.*?) - type: (.*?),.* bsf: 0000:(.*?):`)
+
+	for _, line := range lines {
+		// Matching Regular Expressions
+		matches := pattern.FindStringSubmatch(line)
+
+		// Extract the identifier of the lines starting with 'qat_dev' and confirm the device status is "up"
+		if len(matches) == 4 && !strings.HasSuffix(matches[2], "vf") && strings.Contains(line, "state: up") {
+			qatDev := strings.TrimSpace(matches[1])
+
+			//get the corresponding bsf number
+			bsf := strings.ReplaceAll(matches[3], "0000:", "")
+			availableDev[qatDev] = bsf
+		}
+	}
+
+	if len(availableDev) == 0 {
+		return nil, fmt.Errorf("unable to find an available QAT device. Please check the status of QAT")
+	}
+	return availableDev, nil
+}
+
+// controlTelemetry turn on/off telemetry
+func controlTelemetry(devicesAddr map[string]interface{}, mode int) error {
+	for qatDev, bsf := range devicesAddr {
+		//The path to control the telemetry switch
+		path := fmt.Sprintf(controlPath, bsf, bsf)
+
+		file, err := os.OpenFile(path, os.O_WRONLY, 0644)
+		if err != nil {
+			return fmt.Errorf("failed to open telemetry control file: %s ", err)
+		}
+		defer file.Close()
+
+		//turn on/off telemetry
+		_, err = file.WriteString(strconv.Itoa(mode))
+		if err != nil {
+			return fmt.Errorf("failed to control %s with mode %d: %s ", qatDev, mode, err)
+		}
+	}
+
+	return nil
+}
+
+// getUtilization calculate utilization from each qat device
+func getUtilization(bsf string) (DeviceUtilizationSample, error) {
+
+	utilizationSample := DeviceUtilizationSample{}
+	//dataPath that can read data from telemetry
+	dataPath := fmt.Sprintf(deviceDataPath, bsf, bsf)
+
+	f, err := os.OpenFile(dataPath, os.O_RDONLY, 0666)
```

Comment:
can the data file keep open in `Init()` and close in `Shutdown()`? Just to save some file open overhead.